### PR TITLE
Building Road Intersection Check Enhancement

### PR DESCRIPTION
**Pull request by sayas01**
### Description:

This PR adds following enhancements to the BuildingRoadIntersectionCheck:
1. Adds a configurable `car.navigable` to specify valid highway tags for intersection. The default value of "car.navigable" is set to true. If set to true, the intersecting highways will be checked for tags in the "CAR_NAVIGABLE_HIGHWAYS" enum set of `HighwayTag` class. If set to false, it will be checked for tags in "CORE_WAYS" enum set.
2. Ignores intersections with intersecting highway containing nodes with BarrierTag as these are valid intersections.
3. Ignores intersections with intersecting building containing points with `amenity=fuel` tag. Previously, buildings with `amenity=fuel` was ignored. There are cases where nodes within buildings, which are ingested into Atlas as points, have `amenity=fuel` tag. These are valid intersections and should not be flagged. The enhancement also captures this case.

### Potential Impact:

The enhancements will reduce a large number of false positives for this check.

### Unit Test Approach:

Added three test cases to test the enhancements.

### Test Results:

NA

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 1]
```diff
@@ -29,16 +31,25 @@
...
 public class BuildingRoadIntersectionCheck extends BaseCheck<Long>
 {
+    private static final String BUILDING_ROAD_INTERSECTION_INSTRUCTION = "Building (id-{0,number,#}) intersects road (id-{1,number,#})";
+    private static final String SERVICE_ROAD_INTERSECTION_INSTRUCTION = "Building (id-{0,number,#}) intersects road (id-{1,number,#}), which is a SERVICE road. Please verify whether the intersection is valid or not.";
     private static final List<String> FALLBACK_INSTRUCTIONS = Arrays
-            .asList("Building (id-{0,number,#}) intersects road (id-{1,number,#})");
+            .asList(BUILDING_ROAD_INTERSECTION_INSTRUCTION, SERVICE_ROAD_INTERSECTION_INSTRUCTION);
     private static final String INDOOR_KEY = "indoor";
     private static final String YES_VALUE = "yes";
+    private static final Predicate<Edge> highwayServiceTag = edge -> Validators.isOfType(edge,
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: Can you snake case + caps lock this to fit wit the style for static final constants.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 2]
```diff
@@ -48,29 +59,41 @@
...
+                        || Validators.isOfType(node, AmenityTag.class, AmenityTag.PARKING_ENTRANCE))
+                // Ignore edges with nodes containing Barrier tags
+                || edge.getAtlas()
+                        .nodesWithin(edge.bounds(),
+                                node -> Validators.hasValuesFor(node, BarrierTag.class))
+                        .iterator().hasNext());
     }
 
-    private static Predicate<Edge> intersectsCoreWayInvalidly(final Area building)
+    private Predicate<Edge> intersectsCoreWayInvalidly(final Area building)
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: wherever you use this make sure to put a ```this.``` in front of it.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 3]
```diff
@@ -162,13 +180,14 @@ private void handleIntersections(final Iterable<Edge> intersectingEdges, final C
...
+                final int instructionIndex = highwayServiceTag.test(edge) ? 1 : 0;
+                flag.addObject(edge, this.getLocalizedInstruction(instructionIndex,
+                        building.getOsmIdentifier(), edge.getOsmIdentifier()));
                 knownIntersections.add(edge);
-                if (edge.hasReverseEdge())
+                final Optional<Edge> reverseEdge = edge.reversed();
+                if (reverseEdge.isPresent())
                 {
-                    knownIntersections.add(edge.reversed().get());
+                    knownIntersections.add(reverseEdge.get());
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: for sonar or IDE warning?

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 4]
```diff
@@ -104,6 +127,7 @@ private static boolean isValidIntersection(final Area building, final Edge edge)
     public BuildingRoadIntersectionCheck(final Configuration configuration)
     {
         super(configuration);
+        this.isCarNavigable = this.configurationValue(configuration, "car.navigable", true);
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: Can you refactor this variable to something like ```carNavigableEdgesOnly``` so its a bit more clear.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 5]
```diff
@@ -48,29 +59,41 @@
...
+                        point -> Validators.isOfType(point, AmenityTag.class, AmenityTag.FUEL))
+                        .iterator().hasNext()
                 // And if the layers have the same layer value
                 && LayerTag.getTaggedValue(edge).orElse(0L)
                         .equals(LayerTag.getTaggedValue(building).orElse(0L))
                 // And if the building/edge intersection is not valid
-                && !isValidIntersection(building, edge);
+                && !isValidIntersection(building, edge)
+                // And if the edge has no Access = Private tag
+                && !AccessTag.isPrivate(edge);
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: are private edges allowed to intersect?

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 6]
```diff
@@ -48,29 +59,41 @@
...
+                : HighwayTag.isCoreWay(edge))
                 // And if the edge intersects the building polygon
                 && edge.asPolyLine().intersects(building.asPolygon())
                 // And ignore intersections where edge has highway=service and building has
                 // Amenity=fuel
-                && !(Validators.isOfType(edge, HighwayTag.class, HighwayTag.SERVICE)
+                && !(highwayServiceTag.test(edge)
                         && Validators.isOfType(building, AmenityTag.class, AmenityTag.FUEL))
+                // And ignore intersections where building has points within it with Amenity=fuel
+                && !edge.getAtlas().pointsWithin(building.asPolygon(),
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: so you can do this processing outside of the lambda so you don't do this for every edge. 
```
if( ! building.getAtlas().pointsWithin ....... hasNext())
{
 //rest of logic
}else
{
return ture;
}
```

Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: You are right! As this condition is checked against buildings, I have moved this to the `validCheckForObject()`, which filters valid buildings for the Check.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java (hunk 5 above)
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: Edges with `access=private` was creating a large number of false positives. I have updated this to check only `access=private`.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 7]
```diff
@@ -115,26 +135,24 @@ public boolean validCheckForObject(final AtlasObject object)
         return object instanceof Area && BuildingTag.isBuilding(object)
                 && !HighwayTag.isHighwayArea(object)
                 && !Validators.isOfType(object, AmenityTag.class, AmenityTag.PARKING)
-                && !Validators.isOfType(object, BuildingTag.class, BuildingTag.ROOF);
+                && !Validators.isOfType(object, BuildingTag.class, BuildingTag.ROOF)
+                // Ignore buildings that have points withing it with Ameniity=Fuel
+                && !object.getAtlas().pointsWithin(((Area) object).asPolygon(),
```
Comment (jklamer) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: 👍

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 8]
```diff
@@ -48,29 +59,37 @@
...
-                || edge.connectedNodes().stream().anyMatch(
-                        node -> Validators.isOfType(node, EntranceTag.class, EntranceTag.YES)
-                                || Validators.isOfType(node, AmenityTag.class,
-                                        AmenityTag.PARKING_ENTRANCE)));
+                || edge.connectedNodes().stream().anyMatch(node -> Validators.isOfType(node,
+                        EntranceTag.class, EntranceTag.YES)
+                        || Validators.isOfType(node, AmenityTag.class, AmenityTag.PARKING_ENTRANCE))
+                // Ignore edges with nodes containing Barrier tags
+                || edge.getAtlas()
+                        .nodesWithin(edge.bounds(),
```
[the remark this answers is not in the document]
Comment (hallahan) [reply src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: Also, this lookup is more expensive.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 9]
```diff
@@ -48,29 +59,37 @@
...
+        // An invalid intersection is determined by checking that its highway tag is car navigable
+        // or core way based on the configuration value
+        return edge -> (this.carNavigableEdgesOnly ? HighwayTag.isCarNavigableHighway(edge)
+                : HighwayTag.isCoreWay(edge))
                 // And if the edge intersects the building polygon
                 && edge.asPolyLine().intersects(building.asPolygon())
                 // And ignore intersections where edge has highway=service and building has
                 // Amenity=fuel
-                && !(Validators.isOfType(edge, HighwayTag.class, HighwayTag.SERVICE)
+                && !(HIGHWAY_SERVICE_TAG.test(edge)
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: I like this pattern.

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java [hunk 10]
```diff
@@ -162,13 +180,14 @@ private void handleIntersections(final Iterable<Edge> intersectingEdges, final C
...
             {
-                flag.addObject(edge, this.getLocalizedInstruction(0, building.getOsmIdentifier(),
-                        edge.getOsmIdentifier()));
-
+                final int instructionIndex = HIGHWAY_SERVICE_TAG.test(edge) ? 1 : 0;
+                flag.addObject(edge, this.getLocalizedInstruction(instructionIndex,
+                        building.getOsmIdentifier(), edge.getOsmIdentifier()));
                 knownIntersections.add(edge);
-                if (edge.hasReverseEdge())
+                final Optional<Edge> reverseEdge = edge.reversed();
```
Comment (hallahan) [inline src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: reverseEdge should be reversedEdge

File: src/test/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheckTest.java [hunk 11]
```diff
@@ -9,7 +9,7 @@
 /**
  * {@link BuildingRoadIntersectionCheck} unit test
  *
- * @author mgostintsev
+ * @author mgostintsev, sayas01
```
Comment (hallahan) [inline src/test/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheckTest.java]: Usually you do a new line.

@author mgostintsev
@author sayas01

File: src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sayas01) [reply src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BuildingRoadIntersectionCheck.java]: This has been addressed. Checking the start and end nodes would be sufficient.

**Pull request merged**